File: src/icicl/retriever.py
"""Trajectory retriever for in-context learning."""

from icicl.database import TrajectoryDatabase
from icicl.models import StepExample, Trajectory


class TrajectoryRetriever:
    """Retriever for finding relevant step examples for in-context learning.

    Following the paper, retrieves different examples at each decision point
    based on relevance to the current situation. Uses step-level retrieval
    for fine-grained similarity matching.
    """

    def __init__(self, database: TrajectoryDatabase, k: int = 2) -> None:
        """Initialize the retriever.

        Args:
            database: The trajectory database to search.
            k: Default number of examples to retrieve.
        """
        self._database = database
        self._k = k
        self._retrieved_ids: list[str] = []

    def retrieve_for_plan(self, goal: str, k: int | None = None) -> list[StepExample]:
        """Retrieve step examples for planning phase.

        Args:
            goal: The goal description.
            k: Number of examples to retrieve. Uses default if None.

        Returns:
            List of relevant step examples.
        """
        k = k or self._k
        steps = self._database.search_steps(goal, k=k)
        self._track_retrieved_steps(steps)
        return steps

    def retrieve_for_step(
        self,
        goal: str,
        plan: str,
        observation: str,
        k: int | None = None,
    ) -> list[StepExample]:
        """Retrieve step examples for a reasoning/acting step.

        Args:
            goal: The goal description.
            plan: The current plan.
            observation: The current observation.
            k: Number of examples to retrieve. Uses default if None.

        Returns:
            List of relevant step examples.
        """
        k = k or self._k
        # Query based on current observation for step-level similarity
        query = f"{observation}"
        steps = self._database.search_steps(query, k=k)
        self._track_retrieved_steps(steps)
        return steps

    def _track_retrieved_steps(self, steps: list[StepExample]) -> None:
        """Track which trajectories were retrieved for later curation."""
        for step in steps:
            if step.trajectory_id not in self._retrieved_ids:
                self._retrieved_ids.append(step.trajectory_id)

    def _track_retrieved(self, trajectories: list[Trajectory]) -> None:
        """Track which trajectories were retrieved for later curation (legacy)."""
        for traj in trajectories:
            if traj.id not in self._retrieved_ids:
                self._retrieved_ids.append(traj.id)

    def get_retrieved_ids(self) -> list[str]:
        """Get all trajectory IDs retrieved in this session.

        Returns:
            List of trajectory IDs.
        """
        return list(self._retrieved_ids)

    def clear_retrieved(self) -> None:
        """Clear the list of retrieved trajectory IDs."""
        self._retrieved_ids = []

    def record_episode_result(self, success: bool) -> None:
        """Record the result of the episode for curation.

        Args:
            success: Whether the episode was successful.
        """
        if self._retrieved_ids:
            self._database.record_retrieval(self._retrieved_ids, success)
        self.clear_retrieved()

File: icrl-ts/src/retriever.ts
/**
 * Trajectory retriever for in-context learning.
 */

import type { StepExample } from "./models";
import type { TrajectoryDatabase } from "./database";

/**
 * Retriever for finding relevant step examples for in-context learning.
 *
 * Following the paper, retrieves different examples at each decision point
 * based on relevance to the current situation. Uses step-level retrieval
 * for fine-grained similarity matching.
 */
export class TrajectoryRetriever {
  private readonly database: TrajectoryDatabase;
  private readonly k: number;
  private retrievedIds: string[] = [];

  /**
   * Initialize the retriever.
   *
   * @param database - The trajectory database to search.
   * @param k - Default number of examples to retrieve.
   */
  constructor(database: TrajectoryDatabase, k: number = 2) {
    this.database = database;
    this.k = k;
  }

  /**
   * Retrieve step examples for planning phase.
   *
   * @param goal - The goal description.
   * @param k - Number of examples to retrieve. Uses default if undefined.
   * @returns List of relevant step examples.
   */
  async retrieveForPlan(goal: string, k?: number): Promise<StepExample[]> {
    const numExamples = k ?? this.k;
    const steps = await this.database.searchSteps(goal, numExamples);
    this.trackRetrievedSteps(steps);
    return steps;
  }

  /**
   * Retrieve step examples for a reasoning/acting step.
   *
   * @param goal - The goal description.
   * @param plan - The current plan.
   * @param observation - The current observation.
   * @param k - Number of examples to retrieve. Uses default if undefined.
   * @returns List of relevant step examples.
   */
  async retrieveForStep(
    goal: string,
    _plan: string,
    observation: string,
    k?: number
  ): Promise<StepExample[]> {
    const numExamples = k ?? this.k;
    // Query based on the current task + observation for step-level similarity.
    // (The plan string can be very long/noisy; we focus on goal + observation.)
    const query = `${goal}\n${observation}`;
    const steps = await this.database.searchSteps(query, numExamples);
    this.trackRetrievedSteps(steps);
    return steps;
  }

  private trackRetrievedSteps(steps: StepExample[]): void {
    for (const step of steps) {
      if (!this.retrievedIds.includes(step.trajectoryId)) {
        this.retrievedIds.push(step.trajectoryId);
      }
    }
  }

  /**
   * Get all trajectory IDs retrieved in this session.
   */
  getRetrievedIds(): string[] {
    return [...this.retrievedIds];
  }

  /**
   * Clear the list of retrieved trajectory IDs.
   */
  clearRetrieved(): void {
    this.retrievedIds = [];
  }

  /**
   * Record the result of the episode for curation.
   *
   * @param success - Whether the episode was successful.
   */
  recordEpisodeResult(success: boolean): void {
    if (this.retrievedIds.length > 0) {
      for (const id of this.retrievedIds) {
        this.database.recordRetrieval(id, success);
      }
    }
    this.clearRetrieved();
  }
}
